src/Compiler/Checking/FindUnsolved.fs

```diff
@@ -19,13 +19,42 @@ type env = | NoEnv
 
 let FindUnsolvedStackGuardDepth = StackGuard.GetDepthOption "FindUnsolved"
 
+type SyncStackGuard(maxDepth) =
```

Comment:
Its definitely an elegant way of changing behavior while minimizing risks and diff size.


A downside compared to hand-written, non-generic, usage directly in the processing function (e.g. just storing a `Stack<Expr>`)  is repeated closure for effectively the same argument set.

If that does not turn up being a visible issue, this could be a solution 👍

Reply:
With slight adjustment It is possible to make it work with single closure and a `Stack<Expr>` at the cost of more obscure application, for example:

```fsharp
SyncStackGuard.Guard cenv.stackGuard expr <| fun expr ->
```

instead of current, nicer

```fsharp
cenv.stackGuard.Guard <| fun () ->
```

Reply:
Yes, a closure would still be there, just without the surrounding params.
In that case I think the proposed version with same API, i.e. `cenv.stackGuard.Guard <| fun () ->` is OK 👍